metrics/cadvisor.go

```diff
@@ -107,40 +149,140 @@ func DecodeCAdvisor(r io.Reader) (CadvisorMetrics, error) {
-	// container name is empty if not existent
-	containerName, _ = t.Tags["container_name"]
-	id, ok := t.Tags["id"]
-	if !ok {
-		return
-	}
-	if !strings.HasPrefix(id, "/kubepods") {
-		return
+func (cAdvisor *CAdvisor) GetRawMetrics() (RawMetrics, error) {
+	mutex := sync.Mutex{}
+	rawMetrics := RawMetrics{}
+
+	getNodeMetrics := func(node kuber.Node) error {
+		cAdvisor.Infof(
+			nil,
+			"{cAdvisor} requesting metrics from node %s",
+			node.Name,
+		)
+
+		return cAdvisor.withBackoff(func() error {
+			cAdvisorResponse, err := http.Get(
+				"http://" + cAdvisor.getNodeKubeletAddress(node) + "/metrics/cAdvisorMetrics",
+			)
+			now := time.Now()
+
+			if cAdvisorResponse != nil && cAdvisorResponse.Body != nil {
+				defer func() {
+					if err := cAdvisorResponse.Body.Close(); err != nil {
+						cAdvisor.Errorf(
+							err,
+							"unable to close cAdvisorResponse.Body",
+						)
+					}
+				}()
+			}
+
+			if err != nil {
+				return karma.Format(
+					err,
+					"{cAdvisor} unable to get cAdvisor from node %q",
+					node.Name,
+				)
+			}
+
+			cAdvisorMetrics, err := decodeCAdvisorResponse(cAdvisorResponse.Body)
+
+			if err != nil {
+				return karma.Format(
+					err,
+					"{cAdvisor} unable to decode cAdvisor response from node %q",
+					node.Name,
+				)
+			}
+
+			for metric, values := range cAdvisorMetrics {
+				for _, val := range values {
+
+					var containerName, podName, namespace string
+					var applicationID, serviceID, containerID uuid.UUID
+
+					containerName, _ = val.Tags["container_name"]
+					podName, _ = val.Tags["pod_name"]
+					namespace, _ = val.Tags["namespace"]
+
+					// NOTE: we still add metrics with no bounded entities
+					if containerName != "" {
+						var container *scanner.Container
+						applicationID, serviceID, container, _ = cAdvisor.scanner.FindContainer(namespace, podName,
+							containerName)
+						containerID = container.ID
+					} else if podName != "" {
+						applicationID, serviceID, _ = cAdvisor.scanner.FindService(namespace, podName)
+					}
+
+					var applicationTag, serviceTag, containerTag *uuid.UUID
+					if !applicationID.IsNil() {
+						applicationTag = &applicationID
+					}
+					if !serviceID.IsNil() {
+						serviceTag = &serviceID
+					}
+					if !containerID.IsNil() {
+						containerTag = &containerID
+					}
+
+					mutex.Lock()
+					rawMetrics = append(rawMetrics, &RawMetric{
+						Metric: metric,
+
+						Node: node.ID,
+
+						Application: applicationTag,
+						Service:     serviceTag,
+						Container:   containerTag,
+
+						Tags:  val.Tags,
+						Value: val.Value,
+
+						Timestamp: now,
+					})
+					mutex.Unlock()
+				}
+			}
+
+			return nil
+
+		})
 	}
-	podregexp := regexp.MustCompile(`pod[0-9a-f\-]+`)
-	podUID = podregexp.FindString(id)[3:]
-	if len(podUID) <= 0 {
-		return
+
+	nodes := cAdvisor.scanner.GetNodes()
+	pr, err := alltogether.NewConcurrentProcessor(nodes, getNodeMetrics)
```

Comment:
@YasserRabee there is no need to use alltogether, let's use https://godoc.org/golang.org/x/sync/errgroup instead

Reply:
@mie00 thanks. I will replace alltogether with errgroup. :+1: